### PR TITLE
CDAP-2105: Queue introspection

### PR DESCRIPTION
**Pull request by jwang47**
https://issues.cask.co/browse/CDAP-2105
http://builds.cask.co/browse/CDAP-DUT1461

After beaming CDAP to a cluster, you can run the tool using the following java command from `/opt/cdap/master`:

```
java -cp `hbase classpath`:"lib/*:../hbase-compat-0.98/lib/*":/etc/cdap/conf/* co.cask.cdap.data2.transaction.queue.hbase.HBaseQueueDebugger queue:///default/PurchaseHistory/PurchaseFlow/reader/queue
```

### CODE REVIEW

**Comment by sreevatsanraman** [comment]
Can you add an option to run this from CLI? We should hide this from auto complete so that people don't run this by mistake.

**Comment by jwang47** [comment]
I don't think we should do that, since then cdap-cli would depend on cdap-data-fabric.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+    byte[] stateColumnName = Bytes.add(QueueEntryRow.STATE_COLUMN_PREFIX,
+                                       Bytes.toBytes(consumerConfig.getGroupId()));
+
+
+    int distributorBuckets = queueClientFactory.getDistributorBuckets(hTable.getTableDescriptor());
+    ShardedHBaseQueueStrategy queueStrategy = new ShardedHBaseQueueStrategy(distributorBuckets);
+
+    Scan scan = new Scan();
+    scan.setStartRow(start.getStartRow());
```
Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseQueueDebugger.java]: Working on fixing this part.

**Comment by jwang47** [comment]
Tested and working in distributed cluster. Ran from `/opt/cdap/master` and used the the following command to run the tool:

```
java -cp `hbase classpath`:"lib/*:../hbase-compat-0.98/lib/*":/etc/cdap/conf/* co.cask.cdap.data2.transaction.queue.hbase.HBaseQueueDebugger queue:///default/PurchaseHistory/PurchaseFlow/reader/queue
```

File: cdap-common/src/main/java/co/cask/cdap/common/collect/AllPairCollector.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+package co.cask.cdap.common.collect;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Multimap;
+
+import java.util.Map;
+
+/**
+ * This collector will collect until it runs out of memory, it
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/collect/AllPairCollector.java]: Runs out of memory? Why?

File: cdap-common/src/main/java/co/cask/cdap/common/collect/PairCollector.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ * This can be used to collect with different strategies while iterating
+ * over a stream of elements. For every element in the stream, add the
+ * element to the collector. The collector then indicates whether more
+ * elements are needed (for instance, to collect the first N elements only,
+ * use a collector that returns false after the Nth element has been added.
+ *
+ * @param <KEY> Type of key.
+ * @param <VALUE> Type of value.
+ */
+public interface PairCollector<KEY, VALUE> {
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/collect/PairCollector.java]: Why need this? Can't use the `Collector<Map.Entry<Key, Value>>`?

Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/collect/PairCollector.java]: or at least extend from that one and provide an addition method for putting entries into map.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataFabricDistributedModule.java [hunk 4]
```diff
@@ -63,44 +50,4 @@ public void configure() {
     install(new TransactionModules().getDistributedModules());
 
   }
-
-  /**
-   * Provides implementation of {@link ThriftClientProvider} based on configuration.
-   */
-  @Singleton
-  private static final class ThriftClientProviderSupplier implements Provider<ThriftClientProvider> {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/runtime/DataFabricDistributedModule.java]: Can you use `TransactionClientModule` from Tephra instead? We don't need to change this module now. For the tool purpose, doesn't need to choose client based on cConf.

**Comment by chtyim** [comment]
Need time to review it in details. Probably do it on Monday.

**Comment by sreevatsanraman** [comment]
We will close this and open against 3.0. Since this is a tool and not a part of code. It can come in after code freeze.

**Pull request closed without merging**